app/src/main/java/ca/bc/gov/health/ircreader/utils/PayLoadProcessor.kt

```diff
@@ -1,56 +1,104 @@
 import ca.bc.gov.health.ircreader.model.SHCData
+import com.google.gson.Gson
 
 class PayLoadProcessor {
 
-    /*fun processPayLoad(shcData: SHCData): Int {
+    companion object {
+        const val IMMUNIZATION = "Immunization"
+        const val PATIENT = "Patient"
+        const val JANSSEN_SNOWMED = "28951000087107" // TODO: 03/09/21 This will be removed in future 
+        const val JANSSEN_CVX = "212"
+    }
+
+    enum class ImmuStatus {
+        FULLY_IMMUNIZED,
+        PARTIALLY_IMMUNIZED,
+        NO_RECORD;
+    }
+
+    /*
+    * Below code snippet provides immunization status
+    * No immunization entry is no records
+    * Jannssen 1 dose is fully immunized
+    * Other vaccines 2 doses is fully immunized
+    * Other vaccines 1 dose is partially immunized
+    * */
+    fun fetchImmuStatus(shcData: SHCData): ImmuStatus {
         println(Gson().toJson(shcData))
 
         var entries: List<Entry> = mutableListOf()
         entries = shcData.payload.vc.credentialSubject.fhirBundle.entry
 
-        var code: String?
-        var occurrenceDateTime: String = ""
+        // CVX code and number of immunization entries will contribute in deciding the immunization logic
+        var code = ""
+        var numberOfImmuEntries = 0
+
+        //Iterating through the entries to get code and number of immunization entries
         entries.forEach { entry ->
-            if (entry.resource.resourceType.contentEquals("Immunization")) {
-                code = entry.resource.vaccineCode?.coding?.get(0)?.code
-                if (occurrenceDateTime.isEmpty()
-                    || LocalDate.parse(occurrenceDateTime)
-                        .isBefore(LocalDate.parse(entry.resource.occurrenceDateTime.toString()))
-                )
-                    occurrenceDateTime = entry.resource.occurrenceDateTime.toString()
+            try {
+                if (entry.resource.resourceType.contentEquals(IMMUNIZATION)) {
+                    code = entry.resource.vaccineCode?.coding?.get(0)?.code.toString()
+                    numberOfImmuEntries += 1
+                }
+            } catch (e: Exception) {
+                e.printStackTrace()
             }
-            // TODO: 01/09/21 In progress
-            if (entry.resource.resourceType.contentEquals("Patient")) {
+        }
+
+        if (entries.isEmpty()) {
+            return ImmuStatus.NO_RECORD
+        }
+
+        if (code.isEmpty()) {
+            return ImmuStatus.NO_RECORD
+        }
+
+        //If the vaccine is Janssen then only one dose is required
+        // TODO: 02/09/21 only CVX will be considered in future
+        if (code.contentEquals(JANSSEN_SNOWMED) || code.contentEquals(JANSSEN_CVX)) {
+            return ImmuStatus.FULLY_IMMUNIZED
+        }
 
+        //If other vaccines then 2 doses are required
+        // TODO: 02/09/21 only CVX will be considered in future
+        if (!code.contentEquals(JANSSEN_SNOWMED) && !code.contentEquals(JANSSEN_CVX)) {
```

Comment:
While this will probably work for our specific case, the logic in general is not correct. What if I had one JANSSEN then one PFIZER? I _should_ be fully immunized but this will give me NO_RECORD.

You want something more like this:
```
val oneDoseVaccines = entries.filter { entry -> setOf(JANSSEN_SNOWMED,  JANSSEN_CVX ).contains(entry.resource.vaccineCode?.coding?.get(0) }
if (oneDoseVaccines.length > 0 || entries.length > 1) {
  return ImmuStatus.FULLY_IMMUNIZED;
} else if (entries.length > 0) {
  return ImmuStatus.PARTIALLY_IMMUNIZED;
} else {
  return ImmuStatus.NO_RECORD;
}
```

Reply:
Oh I forgot to filter down to only immunizations first 🤦 

So more like:
```
val vaccines = entries.filter { entry -> entry.resource.resourceType.contentEquals(IMMUNIZATION) }
val oneDoseVaccines = vaccines.filter { entry -> setOf(JANSSEN_SNOWMED,  JANSSEN_CVX ).contains(entry.resource.vaccineCode?.coding?.get(0) }
if (oneDoseVaccines.length > 0 || vaccines.length > 1) {
  return ImmuStatus.FULLY_IMMUNIZED;
} else if (vaccines.length > 0) {
  return ImmuStatus.PARTIALLY_IMMUNIZED;
} else {
  return ImmuStatus.NO_RECORD;
}
```